File: internal/graphqlfile/graphql.go
// Package graphqlfile parses .graphql files and builds a JSON body suitable for
// POSTing to a GraphQL endpoint.
//
// A .graphql file may contain just a GraphQL document (query/mutation), or a
// document preceded by headers:
//
//	# URL: https://api.example.com/graphql
//	# Header: Authorization: Bearer {{TOKEN}}
//	---
//	query UserById($id: ID!) {
//	  user(id: $id) { name email }
//	}
//	---
//	{ "id": "42" }
//
// The first `---` separator splits metadata from the operation; the second
// separates the operation from a JSON variables block. Both separators are
// optional.
package graphqlfile

import (
	"encoding/json"
	"os"
	"strings"

	"github.com/bgunnarsson/binman/internal/httpfile"
)

// Load reads a .graphql file and returns a Request whose body is the
// application/json payload GraphQL servers expect.
func Load(path string) (*httpfile.Request, error) {
	data, err := os.ReadFile(path)
	if err != nil {
		return nil, err
	}
	return Parse(string(data))
}

// Parse parses the text of a .graphql file.
func Parse(content string) (*httpfile.Request, error) {
	req := &httpfile.Request{
		Method:  "POST",
		Headers: map[string]string{"Content-Type": "application/json"},
		RawText: content,
	}

	parts := strings.Split(content, "\n---\n")
	var metaLines []string
	var query string
	var varsJSON string

	switch len(parts) {
	case 1:
		query = parts[0]
	case 2:
		metaLines = strings.Split(parts[0], "\n")
		query = parts[1]
	default:
		metaLines = strings.Split(parts[0], "\n")
		query = parts[1]
		varsJSON = strings.TrimSpace(strings.Join(parts[2:], "\n---\n"))
	}

	for _, line := range metaLines {
		line = strings.TrimSpace(line)
		line = strings.TrimPrefix(line, "#")
		line = strings.TrimSpace(line)
		if line == "" {
			continue
		}
		if v, ok := trimPrefixFold(line, "url:"); ok {
			req.URL = strings.TrimSpace(v)
			continue
		}
		if v, ok := trimPrefixFold(line, "header:"); ok {
			if idx := strings.Index(v, ":"); idx > 0 {
				req.Headers[strings.TrimSpace(v[:idx])] = strings.TrimSpace(v[idx+1:])
			}
		}
	}

	payload := map[string]any{"query": strings.TrimSpace(query)}
	if varsJSON != "" {
		var vars any
		if err := json.Unmarshal([]byte(varsJSON), &vars); err == nil {
			payload["variables"] = vars
		}
	}
	bodyBytes, err := json.Marshal(payload)
	if err != nil {
		return nil, err
	}
	req.Body = string(bodyBytes)
	return req, nil
}

func trimPrefixFold(s, prefix string) (string, bool) {
	if len(s) >= len(prefix) && strings.EqualFold(s[:len(prefix)], prefix) {
		return s[len(prefix):], true
	}
	return "", false
}

File: crates/binman-core/src/formats/graphql.rs
//! `.graphql` files, posted as the JSON document GraphQL servers expect.
//!
//! A file may be the operation alone, or the operation between a header and a
//! variables block:
//!
//! ```text
//! # URL: https://api.example.com/graphql
//! # Header: Authorization: Bearer {{TOKEN}}
//! ---
//! query UserById($id: ID!) {
//!   user(id: $id) { name email }
//! }
//! ---
//! { "id": "42" }
//! ```

use serde_json::Value;

use crate::body::BodyKind;
use crate::request::{Request, set_header};

pub fn parse(text: &str) -> Request {
    let text = text.replace("\r\n", "\n");
    let parts: Vec<&str> = text.split("\n---\n").collect();
    let (meta, query, variables) = match parts.as_slice() {
        [] => (None, "", None),
        [query] => (None, *query, None),
        [meta, query] => (Some(*meta), *query, None),
        [meta, query, rest @ ..] => (Some(*meta), *query, Some(rest.join("\n---\n"))),
    };

    let mut request = Request {
        method: "POST".into(),
        headers: vec![("Content-Type".into(), "application/json".into())],
        kind: Some(BodyKind::Json),
        ..Request::default()
    };

    for line in meta.into_iter().flat_map(str::lines) {
        let line = line.trim().trim_start_matches('#').trim();
        if let Some(url) = strip_prefix_ignoring_case(line, "url:") {
            request.url = url.trim().to_string();
        } else if let Some(header) = strip_prefix_ignoring_case(line, "header:")
            && let Some((name, value)) = header.split_once(':')
            && !name.trim().is_empty()
        {
            set_header(&mut request.headers, name.trim(), value.trim());
        }
    }

    let mut payload = serde_json::Map::new();
    payload.insert("query".into(), Value::String(query.trim().to_string()));
    if let Some(variables) = variables
        .as_deref()
        .map(str::trim)
        .filter(|variables| !variables.is_empty())
        .and_then(|variables| serde_json::from_str::<Value>(variables).ok())
    {
        payload.insert("variables".into(), variables);
    }
    request.body = Value::Object(payload).to_string();
    request
}

fn strip_prefix_ignoring_case<'a>(text: &'a str, prefix: &str) -> Option<&'a str> {
    let head = text.get(..prefix.len())?;
    head.eq_ignore_ascii_case(prefix)
        .then(|| &text[prefix.len()..])
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn an_operation_alone_is_posted_as_json() {
        let request = parse("query { me { id } }");
        assert_eq!(request.method, "POST");
        assert_eq!(request.header("Content-Type"), Some("application/json"));
        let payload: Value = serde_json::from_str(&request.body).unwrap();
        assert_eq!(payload["query"], "query { me { id } }");
    }

    #[test]
    fn reads_the_header_and_the_variables() {
        let request = parse(
            "# URL: https://api.example.com/graphql\n\
             # Header: Authorization: Bearer abc\n\
             ---\n\
             query UserById($id: ID!) {\n  user(id: $id) { name }\n}\n\
             ---\n\
             {\"id\": \"42\"}\n",
        );
        assert_eq!(request.url, "https://api.example.com/graphql");
        assert_eq!(request.header("Authorization"), Some("Bearer abc"));
        let payload: Value = serde_json::from_str(&request.body).unwrap();
        assert_eq!(payload["variables"]["id"], "42");
    }
}
